File: src/main/java/trap/report/ExcelHelper.java
package trap.report;

import org.apache.poi.ss.usermodel.Cell;
import org.apache.poi.ss.usermodel.CellStyle;
import org.apache.poi.ss.usermodel.Row;
import org.apache.poi.ss.usermodel.Sheet;
import org.apache.poi.ss.usermodel.Workbook;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import trap.model.IndividualTotal;
import trap.model.RoundScore;

import java.io.FileOutputStream;
import java.io.IOException;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.Calendar;
import java.util.Date;

public final class ExcelHelper {

    private static final Logger LOGGER = LoggerFactory.getLogger(ExcelHelper.class);
    private static final DateTimeFormatter FILE_NAME_DATE_FORMATTER = DateTimeFormatter.ofPattern("yyyyMMddHHmmss");
    private static final String DEFAULT_FONT_NAME = "Calibri";

    private ExcelHelper() {
        throw new UnsupportedOperationException("This is a utility class and cannot be instantiated");
    }

    public static CellStyle setFontForHeaders(Workbook workbook) {
        var font = workbook.createFont();
        font.setFontName(DEFAULT_FONT_NAME);
        font.setItalic(true);
        font.setBold(true);
        font.setFontHeightInPoints((short) 14);

        var style = workbook.createCellStyle();
        style.setFont(font);
        return style;
    }

    public static CellStyle getCellStyle(Workbook workbook) {
        var mainText = workbook.createFont();
        mainText.setFontName(DEFAULT_FONT_NAME);
        mainText.setFontHeightInPoints((short) 12);

        var mainTextStyle = workbook.createCellStyle();
        mainTextStyle.setFont(mainText);
        return mainTextStyle;
    }

    public static void setCurrentDateHeader(Sheet sheet, String currentDate) {
        sheet.getRow(9).getCell(1).setCellValue(sheet.getRow(9).getCell(1).getStringCellValue() + currentDate);
    }

    public static void setCurrentSeasonHeader(Sheet sheet) {
        var cal = Calendar.getInstance();
        cal.setTime(new Date());
        var month = cal.get(Calendar.MONTH);
        var year = cal.get(Calendar.YEAR);
        var currentSeason = month > 8 ? year + 1 : year;
        sheet.getRow(8).getCell(1).setCellValue(currentSeason + " " + sheet.getRow(8).getCell(1).getStringCellValue());
    }

    public static void createFile(Workbook workbook) throws IOException {
        long start = System.currentTimeMillis();
        String filename = generateFilename();

        LOGGER.info("Creating file: {}", filename);
        try (FileOutputStream fileOutputStream = new FileOutputStream(filename)) {
            LOGGER.info("Writing file...");
            workbook.write(fileOutputStream);
            LOGGER.info("File written successfully.");
        } catch (IOException e) {
            LOGGER.error("Error writing file: {}", e.getMessage());
            throw e;
        }

        LOGGER.info("Created file in {} ms", System.currentTimeMillis() - start);
    }

    private static String generateFilename() {
        var currentDate = LocalDateTime.now();
        return "league-data-" + currentDate.format(FILE_NAME_DATE_FORMATTER) + ".xlsx";
    }

    public static void addCleanData(Row row, RoundScore rowData) {
        row.createCell(0).setCellValue(rowData.eventId());
        row.createCell(1).setCellValue(rowData.event());
        row.createCell(2).setCellValue(rowData.locationId());
        row.createCell(3).setCellValue(rowData.location());
        row.createCell(4).setCellValue(rowData.eventDate());
        row.createCell(5).setCellValue(rowData.squadName());
        row.createCell(6).setCellValue(rowData.team());
        row.createCell(7).setCellValue(rowData.athlete());
        row.createCell(8).setCellValue(rowData.classification());
        row.createCell(9).setCellValue(rowData.gender());
        row.createCell(10).setCellValue(rowData.round1());
        row.createCell(11).setCellValue(rowData.round2());
        row.createCell(12).setCellValue(rowData.round3());
        row.createCell(13).setCellValue(rowData.round4());
        row.createCell(14).setCellValue(rowData.round5());
        row.createCell(15).setCellValue(rowData.round6());
        row.createCell(16).setCellValue(rowData.round7());
        row.createCell(17).setCellValue(rowData.round8());
        row.createCell(18).setCellValue(rowData.type());
    }

    public static void addTeamData(Row row, int startColumn, String team, int total, CellStyle mainTextStyle) {
        createAndStyleCell(row, startColumn, team, mainTextStyle);
        createAndStyleCell(row, startColumn + 1, total, mainTextStyle);
    }

    private static void createAndStyleCell(Row row, int column, Object value, CellStyle style) {
        Cell cell = row.createCell(column);

        if (value instanceof String s) {
            cell.setCellValue(s);
        } else if (value instanceof Integer i) {
            cell.setCellValue(i);
        }

        cell.setCellStyle(style);
    }

    public static void addPlayerData(Row row, int column, String athlete, int total, String team, CellStyle mainTextStyle) {
        if (row == null) {
            throw new IllegalArgumentException("Row cannot be null");
        }

        addTeamData(row, column, athlete, total, mainTextStyle);
        createAndStyleCell(row, column + 2, team, mainTextStyle);
    }

    public static int getRows(Sheet sheet, int rows, IndividualTotal rowData) {
        var row = sheet.createRow(++rows); // Pre-increment to update 'rows' before creating the row
        generateTeamRows(rowData, row);
        return rows;
    }

    public static void generateTeamRows(IndividualTotal rowData, Row row) {
        row.createCell(0).setCellValue(rowData.type());
        row.createCell(1).setCellValue(rowData.team());
        row.createCell(2).setCellValue(rowData.classification());
        row.createCell(3).setCellValue(rowData.athlete());
        row.createCell(4).setCellValue(rowData.total());
    }
}

File: src/main/java/trap/report/ExcelHelper.kt
package trap.report

import lombok.AccessLevel
import lombok.NoArgsConstructor
import org.apache.poi.ss.usermodel.*
import trap.model.IndividualTotal
import trap.model.RoundScore
import java.io.FileOutputStream
import java.io.IOException
import java.text.SimpleDateFormat
import java.util.*

@NoArgsConstructor(access = AccessLevel.PRIVATE)
object ExcelHelper {
    @JvmStatic
    fun setFontForHeaders(workbook: Workbook): CellStyle {
        val font = workbook.createFont()
        font.fontName = "Calibri"
        font.italic = true
        font.bold = true
        font.fontHeightInPoints = 14.toShort()
        val style = workbook.createCellStyle()
        style.setFont(font)
        return style
    }

    @JvmStatic
    fun getCellStyle(workbook: Workbook): CellStyle {
        val mainText = workbook.createFont()
        mainText.fontName = "Calibri"
        mainText.fontHeightInPoints = 12.toShort()
        val mainTextStyle = workbook.createCellStyle()
        mainTextStyle.setFont(mainText)
        return mainTextStyle
    }

    @JvmStatic
    fun setCurrentDateHeader(sheet: Sheet, currentDate: String) {
        sheet.getRow(9).getCell(1).setCellValue(sheet.getRow(9).getCell(1).stringCellValue + currentDate)
    }

    @JvmStatic
    fun setCurrentSeasonHeader(sheet: Sheet) {
        val date = Date()
        val cal = Calendar.getInstance()
        cal.time = date
        val month = cal[Calendar.MONTH]
        val year = cal[Calendar.YEAR]
        val currentSeason = if (month > 8) year + 1 else year
        sheet.getRow(8).getCell(1).setCellValue(currentSeason.toString() + " " + sheet.getRow(8).getCell(1).stringCellValue)
    }

    @JvmStatic
    @Throws(IOException::class)
    fun createFile(workbook: Workbook, filename: String?) {
        val start: Long = System.currentTimeMillis()
        val date = Date()
        val formatter = SimpleDateFormat("yyyyMMddHHmmss")
        val currentDate = formatter.format(date)
        val newFilename = if (filename != null) "$filename.xlsx" else "league-data-$currentDate.xlsx"
        println("Creating file")
        val fileOutputStream = FileOutputStream(newFilename)
        println("Writing file")
        workbook.write(fileOutputStream)
        println("closing output stream")
        fileOutputStream.flush()
        fileOutputStream.close()
        println("Created file $newFilename")
        println("Wrote the contents to a file in " + (System.currentTimeMillis() - start) + "ms")
    }

    @JvmStatic
    fun addCleanData(row: Row, rowData: RoundScore) {
        var cell: Cell = row.createCell(0)
        cell.setCellValue(rowData.eventId.toDouble())
        cell = row.createCell(1)
        cell.setCellValue(rowData.event)
        cell = row.createCell(2)
        cell.setCellValue(rowData.locationId.toDouble())
        cell = row.createCell(3)
        cell.setCellValue(rowData.location)
        cell = row.createCell(4)
        cell.setCellValue(rowData.eventDate)
        cell = row.createCell(5)
        cell.setCellValue(rowData.squadName)
        cell = row.createCell(6)
        cell.setCellValue(rowData.team)
        cell = row.createCell(7)
        cell.setCellValue(rowData.athlete)
        cell = row.createCell(8)
        cell.setCellValue(rowData.classification)
        cell = row.createCell(9)
        cell.setCellValue(rowData.gender)
        cell = row.createCell(10)
        cell.setCellValue(rowData.round1.toDouble())
        cell = row.createCell(11)
        cell.setCellValue(rowData.round2.toDouble())
        cell = row.createCell(12)
        cell.setCellValue(rowData.round3.toDouble())
        cell = row.createCell(13)
        cell.setCellValue(rowData.round4.toDouble())
        cell = row.createCell(14)
        cell.setCellValue(rowData.round5.toDouble())
        cell = row.createCell(15)
        cell.setCellValue(rowData.round6.toDouble())
        cell = row.createCell(16)
        cell.setCellValue(rowData.round7.toDouble())
        cell = row.createCell(17)
        cell.setCellValue(rowData.round8.toDouble())
        cell = row.createCell(18)
        cell.setCellValue(rowData.type)
    }

    @JvmStatic
    fun addTeamData(row: Row, startColumn: Int, team: String?, total: Int, mainTextStyle: CellStyle?) {
        var cell = row.createCell(startColumn)
        cell.setCellValue(team)
        cell.cellStyle = mainTextStyle
        cell = row.createCell(startColumn + 1)
        cell.setCellValue(total.toDouble())
        cell.cellStyle = mainTextStyle
    }

    @JvmStatic
    fun addPlayerData(row: Row?, column: Int, athlete: String?, total: Int, team: String?, mainTextStyle: CellStyle?) {
        if (row != null) {
            var cell = row.createCell(column)
            cell.setCellValue(athlete)
            cell.cellStyle = mainTextStyle
            cell = row.createCell(column + 1)
            cell.setCellValue(total.toDouble())
            cell.cellStyle = mainTextStyle
            cell = row.createCell(column + 2)
            cell.setCellValue(team)
            cell.cellStyle = mainTextStyle
        }
    }

    @JvmStatic
    fun getRows(sheet: Sheet, rows: Int, rowData: IndividualTotal): Int {
        var rows = rows
        val row = sheet.createRow(++rows)
        var cell = row.createCell(0)
        cell.setCellValue(rowData.type)
        cell = row.createCell(1)
        cell.setCellValue(rowData.team)
        cell = row.createCell(2)
        cell.setCellValue(rowData.classification)
        cell = row.createCell(3)
        cell.setCellValue(rowData.athlete)
        cell = row.createCell(4)
        cell.setCellValue(rowData.total.toDouble())
        return rows
    }
}
